File: c_files/raylib/radar_sweep.c
#include <math.h>
#include <raylib.h>

typedef struct Ring
{
  Vector2 center;
  float inner_radius;
  float outer_radius;
  float start_angle;
  float end_angle;
  int segments;
  Color color;

} Ring;

void
ring_init(Ring* ring)
{
  ring->center = (Vector2){ GetScreenWidth() * 0.5f, GetScreenHeight() * 0.5f };
  ring->inner_radius = 250.0f;
  ring->outer_radius = 260.0f;
  ring->start_angle = 0.0f;
  ring->end_angle = 360.0f;
  ring->segments = 64;
  ring->color = BLACK;
}

void
ring_draw(Ring* ring)
{
  DrawRing(ring->center,
           ring->inner_radius,
           ring->outer_radius,
           ring->start_angle,
           ring->end_angle,
           ring->segments,
           ring->color);
}

typedef struct Line
{
  Vector2 start_pos;
  Vector2 end_pos;
  float thickness;
  Color color;

} Line;

void
line_init(Line* line, Ring* ring)
{
  line->start_pos = ring->center;
  line->end_pos = (Vector2){ 0, 0 };
  line->thickness = 10.0f;
  line->color = RED;
}

void
line_draw(Line* line)
{
  DrawLineEx(line->start_pos, line->end_pos, line->thickness, line->color);
}

int
main(void)
{
  const int SCREEN_WIDTH = 800;
  const int SCREEN_HEIGHT = 600;
  const char SCREEN_TITLE[] = "Radar Sweep";
  const Color SCREEN_BACKGROUND = RAYWHITE;

  InitWindow(SCREEN_WIDTH, SCREEN_HEIGHT, SCREEN_TITLE);

  float angle = 0.0f;
  float speed = 50.0f;

  Ring ring;
  Line line;

  ring_init(&ring);
  line_init(&line, &ring);

  while (!WindowShouldClose()) {

    angle += speed * GetFrameTime();
    if (angle >= 360.0f) {
      angle = 0.0f;
    }

    float angle_radians = angle * DEG2RAD;

    line.end_pos.x = cosf(angle_radians) * ring.inner_radius + ring.center.x;
    line.end_pos.y = sinf(angle_radians) * ring.inner_radius + ring.center.y;

    BeginDrawing();
    ClearBackground(SCREEN_BACKGROUND);
    
    ring_draw(&ring);
    line_draw(&line);

    EndDrawing();
  }

  CloseWindow();

  return 0;
}

File: python_files/raylib/single_file_programs/radar_sweep.py
from math import cos, pi, sin

import pyray as p

SCREEN_WIDTH: int = 800
SCREEN_HEIGHT: int = 600
SCREEN_TITLE: str = "Radar Sweep"
SCREEN_BACKGROUND: p.Color = p.RAYWHITE
GAME_FPS: int = 60


class Ring:
    def __init__(self) -> None:
        self.center: p.Vector2 = p.Vector2(
            p.get_screen_width() / 2.0, p.get_screen_height() / 2.0
        )
        self.inner_radius: float = 250.0
        self.outer_radius: float = 260.0
        self.start_angle: float = 0.0
        self.end_angle: float = 360.0
        self.segments: int = 64
        self.color: p.Color = p.DARKGREEN

    def draw(self) -> None:
        p.draw_ring(
            self.center,
            self.inner_radius,
            self.outer_radius,
            self.start_angle,
            self.end_angle,
            self.segments,
            self.color,
        )


class Line:
    def __init__(self, ring_center: p.Vector2) -> None:
        self.start_pos: p.Vector2 = p.Vector2(ring_center.x, ring_center.y)
        self.end_pos: p.Vector2 = p.Vector2(ring_center.x, ring_center.y)
        self.thickness: float = 5.0
        self.color: p.Color = p.RED

        self.angle_radians: float = 0.0
        self.angular_speed: float = 1.5

    def draw(self) -> None:
        p.draw_line_ex(
            self.start_pos, self.end_pos, self.thickness, self.color
        )

    def update(
        self, dt: float, ring_center: p.Vector2, ring_inner_radius
    ) -> None:
        self.angle_radians += self.angular_speed * dt
        if self.angle_radians > 2 * pi:
            self.angle_radians = 0.0

        self.end_pos.x = ring_center.x + (
            ring_inner_radius * sin(self.angle_radians)
        )
        self.end_pos.y = ring_center.y - (
            ring_inner_radius * cos(self.angle_radians)
        )


def main() -> None:
    p.init_window(SCREEN_WIDTH, SCREEN_HEIGHT, SCREEN_TITLE)
    p.set_target_fps(GAME_FPS)

    ring: Ring = Ring()
    line: Line = Line(ring.center)

    while not p.window_should_close():
        dt: float = p.get_frame_time()

        line.update(dt, ring.center, ring.inner_radius)

        p.begin_drawing()
        p.clear_background(SCREEN_BACKGROUND)

        ring.draw()
        line.draw()

        p.end_drawing()

    p.close_window()


if __name__ == "__main__":
    main()
